AppRouterTests/AppRouterTests.swift

```diff
@@ -9,56 +9,106 @@ import XCTest
-    let urls = RouterPatterns.allPatterns
-    let (sut, proxy) = makeSUT()
+public protocol NavigatorProtocol {
+    func register(_ pattern: URLPattern, _ factory: @escaping ViewControllerFactory)
+}
 
-    urls.forEach { url in
-      sut.setNavigationMap(url: url)
+public class AppRouter {
+    private let navigator: NavigatorProtocol
+    
+    public init(navigator: NavigatorProtocol) {
+        self.navigator = navigator
+        navigator.register("https://icook.tw/amp/categories/<int:id>") { (_, value, _) -> UIViewController? in
+            guard let id = value["id"] as? Int else {
+                return ErrorVC()
+            }
+            let vc = CategoryVC()
+            vc.id = id
+            return vc
+        }
     }
+    
+}
 
-    XCTAssertEqual(proxy.registeredURLs, urls, "Expected proxy successful register all paths.")
-  }
-
-  // MARK: - Helpers
-
-  private func makeSUT(file: StaticString = #file, line: UInt = #line) -> (sut: Router, proxy: NavigatorProxySpy) {
-    let proxy = NavigatorProxySpy()
-    let sut = Router(proxy: proxy)
-    trackForMemoryLeaks(sut)
-    trackForMemoryLeaks(proxy)
-    return (sut, proxy)
-  }
-
-  private func trackForMemoryLeaks(_ instance: AnyObject, file: StaticString = #file, line: UInt = #line) {
-    addTeardownBlock { [weak instance] in
-      XCTAssertNil(instance, "Instance should have been deallocated. Potential memory leak.", file: file, line: line)
+class AppRouterTests: XCTestCase {
+    func test_requestRegisterPattern_onInit() {
+        // given
+        let myPattern = "https://icook.tw/amp/categories/<int:id>"
+        
+        //when
+        let (_, spy) = makeSUT()
+        
+        // then
+        XCTAssertEqual(spy.registeredPattern, myPattern)
     }
-  }
-
-  private class NavigatorProxySpy: NavigatorProxy {
-    var navigator: Navigator = Navigator()
-    var registeredURLs = [String]()
-
-    func register(from path: String, completion: () -> Void) {
-      registeredURLs.append(path)
-      completion()
+    
+    func test_mapToCategoryVC_onNavigatorExtractIntID() {
+        // given
+        let (_, spy) = makeSUT()
+        
+        //when
+        let mappedVC = spy.triggeredToExtract(["id": 55]) as? CategoryVC
+        
+        // then
+        XCTAssertEqual(mappedVC?.id, 55)
     }
-  }
-}
+    
+    func test_mapToCategoryVC_onNavigatorExtractEmptyValue() {
+        // given
+        let (_, spy) = makeSUT()
+        
+        //when
+        let mappedVC = spy.triggeredToExtract([:])
+        
+        // then
+        XCTAssertTrue(mappedVC is ErrorVC)
+    }
+    
+    // MARK: - Helpers
+    private func makeSUT(
+        file: StaticString = #file,
+        line: UInt = #line
+    ) -> (sut: AppRouter, navigatorSpy: NavigatorSpy)
+    {
+        let navigatorSpy = NavigatorSpy()
+        let sut = AppRouter(navigator: navigatorSpy)
+        trackForMemoryLeaks(sut)
+        trackForMemoryLeaks(navigatorSpy)
```

Comment:
TrackForMemoryLeaks can show Failure message inline with caller, make it more readable.
```suggestion
        trackForMemoryLeaks(sut, file: file, line: line)
        trackForMemoryLeaks(navigatorSpy, file: file, line: line)
```